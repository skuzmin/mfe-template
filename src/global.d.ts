declare module '*.scss?inline' {
  const content: string;
  export default content;
}

/// <reference types="vite/client" />
interface ImportMetaEnv {
  readonly VITE_APP_VERSION: string;
  readonly VITE_BACKEND_URL: string;
  readonly VITE_FRONTEND_URL: string;
  readonly VITE_ENV: string;
  readonly VITE_MFE_NAME: string;
}

interface ImportMeta {
  readonly env: ImportMetaEnv;
}

declare const __COMMIT_HASH__: string;
declare const __BUILT_AT__: number;
