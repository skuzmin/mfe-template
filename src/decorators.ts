const observedAttributes: Array<string> = [];
const watchers = new Map<string, string>();

export function Component(constructor: Function): void {
  Object.defineProperty(constructor, 'observedAttributes', {
    get() {
      return observedAttributes;
    },
    configurable: true,
  });

  constructor.prototype.attributeChangedCallback = function (
    name: string,
    oldVal: string,
    newVal: string,
  ) {
    this[name] = newVal;
    const method = watchers.get(name);
    if (method && typeof this[method] === 'function') {
      this[method](oldVal, newVal);
    }
  };
}

export function Attribute(): PropertyDecorator {
  return function (_: unknown, propertyKey: string | symbol) {
    if (
      typeof propertyKey === 'string' &&
      !observedAttributes.includes(propertyKey)
    ) {
      observedAttributes.push(propertyKey);
    }
  };
}

export function Watch(attr: string): MethodDecorator {
  return function (_: unknown, propertyKey: string | symbol) {
    if (typeof propertyKey === 'string') {
      watchers.set(attr, propertyKey);
    }
  };
}
