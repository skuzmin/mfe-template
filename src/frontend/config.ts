export const namespace = import.meta.env.VITE_MFE_NAME;
