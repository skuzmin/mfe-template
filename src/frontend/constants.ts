export const ActionIds = {
  AnotherOneBtn: 'another-one-btn',
} as const;
