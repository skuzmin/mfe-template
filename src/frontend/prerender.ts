import skeletonStyles from './styles/skeleton.scss?inline';

export const Prerender = () => {
  return `
    <div>
        <style>
          ${skeletonStyles}
        </style>
        <div class="skeleton-container">
          <div class="skeleton-item"></div>
          <div class="skeleton-item"></div>
          <div class="skeleton-item"></div>
        </div>
    </div>`;
};
