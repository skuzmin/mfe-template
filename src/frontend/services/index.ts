export * from './api.service';
export * from './elements.service';
