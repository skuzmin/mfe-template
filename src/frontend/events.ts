import { namespace } from './config';
import type { CatResponse } from '../types';

export type EventType = 'ready' | 'loaded';

export type EventCallback<T> = T extends 'ready'
  ? (payload: string) => void
  : T extends 'loaded'
    ? (payload: Array<CatResponse>) => void
    : () => void;

export type EventMap = { [K in EventType]: EventCallback<K> };

const usePublish = <T>(namespace: string, eventType: EventType) => {
  return (host: HTMLElement) => {
    return (payload: T): void => {
      host.dispatchEvent(
        new CustomEvent(`${namespace}.${eventType}`, {
          detail: payload,
          bubbles: true,
          composed: true,
        }),
      );
    };
  };
};

let eventStore: EventMap;
const eventMap = (host: HTMLElement) => ({
  ready: usePublish<string>(namespace, 'ready')(host),
  loaded: usePublish<CatResponse[]>(namespace, 'loaded')(host),
});

export const events = {
  setHost: (host: HTMLElement) => (eventStore = eventMap(host)),
  publish: () => eventStore,
};
