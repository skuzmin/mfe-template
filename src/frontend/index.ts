import style from './styles/styles.scss?inline';
import type { CatResponse, Locale } from '../types';
import { Attribute, Component, Watch } from '../decorators';
import { namespace } from './config';
import { events } from './events';
import { l10n } from '../l10n';
import { Prerender } from './prerender';
import { ActionIds } from './constants';
import { getCats, ElementsService } from './services';

@Component
export class MfeSeed extends HTMLElement {
  @Attribute() locale: Locale = 'en_GB';

  private isReady?: boolean;
  private apiAbortController?: AbortController;
  private data: Array<CatResponse>;

  constructor() {
    super();
    this.attachShadow({ mode: 'open' });
    events.setHost(this);
    l10n.initialize(this.locale);
    ElementsService.init(this.shadowRoot!);
    this.data = [];
    this.isReady = false;
  }

  //#region Init
  async connectedCallback() {
    await this.init();
    this.render();
    this.initEventListeners();
    this.isReady = true;
    events.publish().ready('MFE ready!');
  }

  async init(): Promise<void> {
    this.renderSkeleton();
    await this.loadData();
  }

  async loadData(): Promise<void> {
    if (this.apiAbortController) {
      this.apiAbortController.abort();
    }
    this.apiAbortController = new AbortController();

    try {
      this.data = await getCats(this.apiAbortController);
      events.publish().loaded(this.data);
    } catch {
      this.renderFallback();
    }
  }

  initEventListeners(): void {
    this.shadowRoot!.addEventListener('click', (e: Event) =>
      this.clickEventHandler(e),
    );
    this.shadowRoot!.addEventListener('keypress', (e: Event) => {
      if ((e as KeyboardEvent).key === 'Enter') {
        e.preventDefault();
        e.stopPropagation();
        this.clickEventHandler(e);
      }
    });
  }

  //#endregion

  //#region Render
  renderFallback() {
    this.shadowRoot!.innerHTML = `
        <div class="${namespace}">
          <style>
            ${style}
          </style>
          <div class="fallback-error-message">
            <h2>${l10n.t('general.error')}</h2>
          </div>
        </div>
      `;
  }

  renderSkeleton() {
    this.shadowRoot!.innerHTML = Prerender();
  }

  render(): void {
    this.shadowRoot!.innerHTML = this.template();
  }

  template() {
    return `
        <div class="${namespace}">
            <style>
              ${style}
            </style>
            <h3 class="${namespace}-title">${l10n.t('general.title')}</h3>
            <div class="${namespace}-container">
                <img class="${namespace}-image" src="${this.data[0]?.url}" alt="cat" />
                <button data-action="${ActionIds.AnotherOneBtn}" class="${namespace}-btn">${l10n.t('general.btnText')}</button>
            </div>
        </div>
    `;
  }

  //#endregion

  //#region Event handlers
  clickEventHandler(event: Event): void {
    const { action } = (event.target as HTMLElement).dataset;
    if (!action) {
      return;
    }
    switch (action) {
      case ActionIds.AnotherOneBtn:
        this.getAnotherCat();
        break;
    }
  }

  getAnotherCat(): void {
    this.renderSkeleton();
    this.loadData().then(() => this.render());
  }

  //#endregion

  //#region Watcher
  @Watch('locale')
  updateName(oldValue: string, newValue: string) {
    console.info('updateLocale', oldValue, newValue);
    l10n.initialize(this.locale);
    if (this.isReady) {
      this.render();
    }
  }

  //#endregion
}

if (!customElements.get(namespace)) {
  customElements.define('mfe-seed', MfeSeed);
}

export const prerender = () => Prerender();
