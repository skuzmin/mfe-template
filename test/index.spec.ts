import { beforeAll, describe, expect, it } from 'vitest';
import { namespace } from '../src/frontend/config';
import { waitForSelector } from './utils';
import { MfeSeed } from '../src/frontend';

describe('index.ts', () => {
  beforeAll(async () => {
    if (!customElements.get(namespace)) {
      customElements.define(namespace, MfeSeed);
      await customElements.whenDefined(namespace);
    }
  });

  it('should render correctly', async () => {
    // Arrange
    const element = document.createElement(namespace);
    const shadowRoot = element.shadowRoot as ShadowRoot;
    document.body.append(element);

    // Act
    const foundElement = await waitForSelector(shadowRoot, `.${namespace}`);

    // Assert
    expect(foundElement).toBeInstanceOf(Element);
    expect(shadowRoot.innerHTML).toMatchSnapshot();
  });
});
