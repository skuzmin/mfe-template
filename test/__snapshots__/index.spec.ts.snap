// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`index.ts > should render correctly 1`] = `
"
        <div class="mfe-seed">
            <style>
              
            </style>
            <h3 class="mfe-seed-title">Cat inspector</h3>
            <div class="mfe-seed-container">
                <img class="mfe-seed-image" src="undefined" alt="cat">
                <button data-action="another-one-btn" class="mfe-seed-btn">Another one</button>
            </div>
        </div>
    "
`;
